src/gpuarray_reduction.c

```diff
@@ -6,841 +6,4095 @@
+	axis->len      = len;
+	axis->splitLen = 1;
+	axis->i0S      = 0;
+	
+	axis->s0S      = s0S;
+	axis->d0S      = 0;
+	axis->d1S      = 0;
+}
+
+/**
+ * @brief Mark axis as reduction axis, with position reduxNum in the axis list.
+ */
+
+static void        axisMarkReduced                  (axis_desc*           axis, int    reduxNum){
+	axis->isReduced = 1;
+	axis->reduxNum  = reduxNum;
+}
+
+/**
+ * @brief Mark axis as (split) intrablock axis.
+ */
+
+static void        axisMarkIntraBlock               (axis_desc*           axis,
+                                                     int                  ibNum,
+                                                     size_t               ibLen){
+	axis->isIntra  = 1;
+	axis->ibNum    = ibNum;
+	axis->splitLen = ibLen;
+}
+
+/**
+ * @brief Get properties of an axis.
+ */
+
+static int         axisGetReduxNum                  (const axis_desc*     axis){
+	return axis->reduxNum;
+}
+static size_t      axisGetLen                       (const axis_desc*     axis){
+	return axis->len;
+}
+static size_t      axisGetIntraLen                  (const axis_desc*     axis){
+	if       (axisIsSplit(axis)){
+		return axis->splitLen;
+	}else if (axisIsIntra(axis)){
+		return axis->len;
+	}else{
+		return 1;
+	}
+}
+static size_t      axisGetInterLen                  (const axis_desc*     axis){
+	if       (axisIsSplit(axis)){
+		return DIVIDECEIL(axis->len, axis->splitLen);
+	}else if (axisIsIntra(axis)){
+		return 1;
+	}else{
+		return axis->len;
+	}
+}
+static size_t      axisGetIntraInterLen             (const axis_desc*     axis){
+	return axisGetIntraLen(axis)*axisGetInterLen(axis);
+}
+static ssize_t     axisGetS0Stride                  (const axis_desc*     axis){
+	return axisGetLen(axis) > 1 ? axis->s0S : 0;
+}
+static size_t      axisGetS0AbsStride               (const axis_desc*     axis){
+	return axisGetS0Stride(axis)<0 ? -(size_t)axisGetS0Stride(axis):
+	                                  +(size_t)axisGetS0Stride(axis);
+}
+static ssize_t     axisGetD0Stride                  (const axis_desc*     axis){
+	return axisGetLen(axis) > 1 ? axis->d0S : 0;
+}
+static size_t      axisGetD0AbsStride               (const axis_desc*     axis){
+	return axisGetD0Stride(axis)<0 ? -(size_t)axisGetD0Stride(axis):
+	                                  +(size_t)axisGetD0Stride(axis);
+}
+static ssize_t     axisGetD1Stride                  (const axis_desc*     axis){
+	return axisGetLen(axis) > 1 ? axis->d1S : 0;
+}
+static size_t      axisGetD1AbsStride               (const axis_desc*     axis){
+	return axisGetD1Stride(axis)<0 ? -(size_t)axisGetD1Stride(axis):
+	                                     +(size_t)axisGetD1Stride(axis);
+}
+static size_t      axisGetI0Stride                  (const axis_desc*     axis){
+	return axis->i0S;
+}
+static void        axisSetI0Stride                  (axis_desc*           axis,
+                                                     size_t               i0S){
+	axis->i0S = i0S;
+}
```

Comment:
Some getters and setters are more justified than others, but my code went through several iterations where certain attributes were computed automatically based on others, and this flipped often enough that I made functions to protect the rest of the code from that concern. I'll have to extirpate many of them now that the design is pretty much frozen.

There are also various places in the code I chose to stick with getters mostly because I liked the uniformity of all attribute accesses going through a getter.

Could absolutely be changed.